### PR TITLE
Distributed Machine learning

### PR DESCRIPTION
**Pull request by vidit23**

### CODE REVIEW

File: apr20/vvb238.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+
+## Pre-knowledge
+**Hardware primitives:** It is a process that is composed of several instructions and is used to complete a certain function. Primitive or atomic action is a program composed of several machine instructions to complete a certain function, and it is indivisible. That is, the execution of primitive must be continuous, and it is not allowed to be interrupted during the execution.
+
+**Memory Latency Concealment:** Latency refers to the delay from requesting to receiving the response; memory delay refers to the delay caused by waiting for the completion of the access to the data stored in the system memory. Memory hiding latency is accomplished by concurrent operations of memory reads.
+
+**AllReduce** AllReduce operation expects each participating process to provide an equally-sized tensor, collectively applies a given arithmetic operation (e.g., sum, prod, min, max) to input tensors from all processes, and returns the same result tensor to each participant.
```
Comment (kmish0923) [inline apr20/vvb238.md:20]: Same result or output of the same functions?

File: apr20/vvb238.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+### PyTorch Distributed
+PyTorch builds an autograd graph to record actions performed. Then, in the backward pass, it uses the autograd graph to conduct backpropagation to generate gradients. The purpose of data parallel is to speed up these training steps on large datasets. To ensure the success of PyTorch Distributed, it needs:
+1. Mathematical equivalence - Applications expect to harvest the same result model as if all training had been performed locally without model replication
+2. Non-intrusive and interceptive API - Development usually starts from local models and then scales out when necessary, thus the developer should be able to reuse the local training script with minimal modifications. while exposing as many optimization opportunities as possible to the internal implementation.
```
Comment (kmish0923) [inline apr20/vvb238.md:33]: typo: minimal modifications, while*

File: apr20/vvb238.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+# Trade-Offs
+
+### TVM
+Using TVM only makes sense if the technique is fixed. Currently in the field of Deep Learning there is a lot of innovation going around and new architectures are being experimented with. If after each change in code, we needed to run a Automated optimization system which searches through a huge space (albiet faster than others), it doesnt represent the fast nature of experimenting.
```
Comment (kmish0923) [inline apr20/vvb238.md:85]: Am I understanding this correctly: would it make sense to finalize your model, and then use TVM so that it is compatible with other hardware?

File: apr20/vvb238.md (hunk 1 above)
Comment (vidit23) [reply apr20/vvb238.md:20]: It returns a tensor of the same size by applying operations such as sum. I'll make that more clear in the notes

File: apr20/vvb238.md (hunk 3 above)
Comment (vidit23) [reply apr20/vvb238.md:85]: This is just my conjecture. There weren't explicit tradeoffs mentioned in the paper. Pytorch is pretty quick with the graph optimization but it doesn't need to do the whole creating schedules and searching for the best one. I would assume you would have to wait a long time for your program to start. Their idea is that because training takes such a long time, adding a little bit more shouldn't make a difference but at least with what I am doing in my DL class, I do not agree.

File: apr20/vvb238.md (hunk 1 above)
Comment (kmish0923) [reply apr20/vvb238.md:20]: thanks!

File: apr20/vvb238.md (hunk 3 above)
Comment (kmish0923) [reply apr20/vvb238.md:85]: Ah i see, thanks for explaining!

File: apr20/vvb238.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+1. Collective communication performs poorly on small tensors. This is solved by **Gradient Bucketing**. DDP achieves higher throughput and lower latency if it waits for a short period of time and buckets multiple gradients into one AllReduce operation. However, DDP should not communicate all gradients in one single AllReduce, otherwise, no communication can start before the computation is over. Thus, with relatively small bucket sizes, DDP can launch AllReduce operations concurrently with the backward pass to overlap communication with computation, which would make a difference in per iteration latency. The optimal value lives in around 25 MB and depends on the type of communication backend employed.
```
Comment (neeraj71) [inline apr20/vvb238.md:64]: I was thinking that type of model (computation graph) should dominate in determining bucket size.

Comment (vidit23) [reply apr20/vvb238.md:64]: Yes, the size of the model matters too but I thought about it like this, ideally, we would want to communicate as asynchronously as possible which would mean send gradients as soon as they are computed but this would crowd the network. So determining the correct bucket size to not overcrowd the network is essential and thus depends on the type of communication backend.
